proto/connectrpc/conformance/v2/client_compat.proto

```diff
@@ -73,6 +74,38 @@ message ClientCompatRequest {
   // For client- or bidi-streaming requests, this delay should be
   // applied before each request sent.
   uint32 request_delay_ms = 18;
+  // If present, the client should cancel the RPC instead of
+  // allowing to complete normally.
+  Cancel cancel = 19;
+  message Cancel {
+    // These fields determine the timing of cancellation.
+    // If none are present, the client should cancel immediately
+    // after all request messages are sent the stream half-closed
+    // (as if the after_half_close_ms field were present and zero).
+    oneof cancel_timing {
```

Comment:
Should we also have the ability to specify never half-closing and _only_ cancelling? Not sure if the expected behavior would be any different.

Reply:
My expectation is that really depends on the client library. At a low-level, you should _never_ have to half-close after cancelling: cancelling implicitly closes the whole thing. So that's the behavior I imagined for `before_half_close`. In HTTP/2 terms, once the stream is cancelled, nothing else is ever sent on it. When you half-close a stream, a final frame with "end of stream" bit set is sent to tell the server the stream is half-closed. So half-closing on the client-side shouldn't actually do anything on the wire if cancel actually works.

In practice however, it is possible that a particular implementation may want client code to use half-close/close-send, even after a cancellation, to ensure there are no resource leaks (like if there are other resources associated with the stream that need to be explicitly freed vs. auto-freed on cancellation).

Hopefully for our current implementations, half-close isn't necessary after cancel, and that's how I'd recommend we implement them.